docs/datatypes/stm.md

```diff
@@ -0,0 +1,115 @@
+    _         <- to.update(existing => existing + amount)
+    recvBal   <- to.get
+  } yield recvBal
+
+val program: IO[String, Long] = for {
+  sndAcc  <- STM.atomically(TRef.make(1000L))
+  rcvAcc  <- STM.atomically(TRef.make(0L))
+  recvAmt <- STM.atomically(transferMoney(sndAcc, rcvAcc, 500L))
+} yield recvAmt
+```
+
+`transferMoney` describes an atomic transfer process between a sender and a receiver. The transaction will fail if the
+sender does not have enough of money in their account. This means that individual accounts will be debited and credited
+atomically. If the transaction fails in the middle, the entire process will be rolled back, and it will appear that 
+nothing has happened. Here, we see that `STM` effects compose using a for-comprehension and that wrapping an `STM` effect
+with `STM.atomically` (or calling `commit` on any STM effect) turns the `STM` effect into a `ZIO` effect which can be
+executed. STM transactions compose sequentially. By using `STM.atomically` (or `commit`), the programmer identifies 
+atomic transaction in the sense that the entire set of operations within `STM.atomically` appears to take place 
+indivisibly.
+
+## Exceptions
+
+`STM` supports exceptions just like `ZIO`. In `transferMoney`, we saw an example of an exception (`STM.fail`). 
+Exceptions in `STM` have abort semantics: if an atomic transaction throws an exception, the transaction is rolled back
+with no effect.
+
+## `retry`
+
+`STM.retry` is central to making transactions composable when they may block. For example, if we wanted to ensure that
+the money transfer took place when the sender had enough of money (instead of failing right away), we can use `STM.retry`
+instead:
+
+```scala mdoc:silent
+def transferMoneyNoMatterWhat(from: TRef[Long], to: TRef[Long], amount: Long): STM[String, Long] =
+  for {
+    senderBal <- from.get
+    _         <- if (senderBal < amount) STM.retry else STM.unit
+    _         <- from.update(existing => existing - amount)
+    _         <- to.update(existing => existing + amount)
+    recvBal   <- to.get
+  } yield recvBal
+```
+
+`STM.retry` will abort and retry the entire transaction until it succeeds (instead of failing like the previous example). 
+Note that the transaction will only be retried when one of the underlying transactional data structures have been changed.
+
+## Composing alternatives
+
+STM transactions compose sequentially so that both STM effects are executed. However, STM transactions can also compose 
+transactions as alternatives so that only one STM effect is executed by making use of `orElse` on STM effects. Provided
+we have two STM effects `sA` and `sB`, you can express that you would like to compose the two using `sA orElse sB`. The
+transaction would first attempt to run `sA` and if it retries then `sA` is abandoned with no effect and then `sB` runs.
+Now if `sB` also retries then the entire call retries. However, it waits for the transactional data structures to change 
```

Comment:
Is this correct? I thought there was an issue that if the second effect retries then we only retry the second effect instead of both effects?

Reply:
Hmm, I was expecting the same semantics as the [STM paper](https://simonmar.github.io/bib/papers/stm.pdf) (Section 3.4) as they say that both `s1` and `s2` should be retried if `s1` is retried and `s2` is retried.

Is this how ZIO's implementation is working?

Reply:
Keep the docs as they are,as this is the desired behavior. I will double check its behavior for any issues. @adamgfraser could you please paste the link to the issue?